File: apps/android-pos/app/src/main/java/com/pos1/app/network/ApiClient.kt
package com.pos1.app.network

import com.pos1.app.data.SecurePrefs
import kotlinx.serialization.json.Json
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Unified API client with JWT auth, token refresh, and device auth.
 */
class ApiClient(private val prefs: SecurePrefs) {

    val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
        encodeDefaults = true
    }

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .addInterceptor(AuthInterceptor(prefs))
        .build()

    val http: OkHttpClient get() = httpClient

    private val jsonMediaType = "application/json".toMediaType()

    // ── Auth ──

    fun login(email: String, password: String): JSONObject {
        val body = JSONObject()
            .put("email", email)
            .put("password", password)
            .toString()
        return post("/api/auth/login", body, auth = false)
    }

    fun refreshToken(refreshToken: String): JSONObject {
        val body = JSONObject()
            .put("refreshToken", refreshToken)
            .toString()
        return post("/api/auth/refresh", body, auth = false)
    }

    fun logout(refreshToken: String): JSONObject {
        val body = JSONObject()
            .put("refreshToken", refreshToken)
            .toString()
        return post("/api/auth/logout", body)
    }

    // ── Device Registration ──

    fun registerDevice(branchId: String, code: String): JSONObject {
        val body = JSONObject()
            .put("branchId", branchId)
            .put("code", code)
            .put("name", "Android POS")
            .put("platform", "ANDROID")
            .toString()
        return post("/api/pos/devices/register", body, auth = false)
    }

    fun rotateDeviceToken(deviceId: String, oldToken: String): JSONObject {
        val body = JSONObject()
            .put("deviceId", deviceId)
            .put("oldToken", oldToken)
            .toString()
        return post("/api/pos/devices/rotate-token", body, auth = false)
    }

    // ── Snapshot ──

    fun fetchSnapshot(branchId: String): JSONObject {
        return get("/api/pos/sync/snapshot?branchId=$branchId", useDeviceAuth = true)
    }

    // ── Orders ──

    fun getOrders(branchId: String, status: String? = null, page: Int = 1, limit: Int = 50): JSONObject {
        val params = buildString {
            append("?branchId=$branchId&page=$page&limit=$limit")
            if (!status.isNullOrBlank()) append("&status=$status")
        }
        return get("/api/orders$params")
    }

    fun createOrder(branchId: String, type: String, items: org.json.JSONArray, tableId: String? = null): JSONObject {
        val body = JSONObject()
            .put("branchId", branchId)
            .put("type", type)
            .put("items", items)
        if (!tableId.isNullOrBlank()) body.put("tableId", tableId)
        return post("/api/orders", body.toString())
    }

    fun updateOrderStatus(orderId: String, status: String): JSONObject {
        val body = JSONObject().put("status", status).toString()
        return patch("/api/orders/$orderId/status", body)
    }

    fun voidOrder(orderId: String, reason: String? = null): JSONObject {
        val body = JSONObject()
        if (!reason.isNullOrBlank()) body.put("reason", reason)
        return patch("/api/orders/$orderId/void", body.toString())
    }

    fun refundOrder(orderId: String, paymentId: String, amount: Double, reason: String? = null): JSONObject {
        val body = JSONObject()
            .put("paymentId", paymentId)
            .put("amount", amount)
        if (!reason.isNullOrBlank()) body.put("reason", reason)
        return patch("/api/orders/$orderId/refund", body.toString())
    }

    fun applyDiscount(orderId: String, discountAmount: Double, reason: String? = null): JSONObject {
        val body = JSONObject()
            .put("discountAmount", discountAmount)
        if (!reason.isNullOrBlank()) body.put("reason", reason)
        return patch("/api/orders/$orderId/discount", body.toString())
    }

    fun splitOrder(orderId: String, items: org.json.JSONArray): JSONObject {
        val body = JSONObject().put("items", items).toString()
        return post("/api/orders/$orderId/split", body)
    }

    fun mergeOrders(sourceOrderId: String, targetOrderId: String): JSONObject {
        val body = JSONObject().put("targetOrderId", targetOrderId).toString()
        return post("/api/orders/$sourceOrderId/merge", body)
    }

    fun transferTable(orderId: String, tableId: String, reason: String? = null): JSONObject {
        val body = JSONObject().put("tableId", tableId)
        if (!reason.isNullOrBlank()) body.put("reason", reason)
        return patch("/api/orders/$orderId/transfer-table", body.toString())
    }

    fun getInvoice(orderId: String): ByteArray? {
        val request = Request.Builder()
            .url("${prefs.apiBaseUrl}/api/orders/$orderId/invoice")
            .get()
            .build()
        httpClient.newCall(request).execute().use { resp ->
            if (!resp.isSuccessful) return null
            return resp.body?.bytes()
        }
    }

    // ── Coupons ──

    fun validateCoupon(code: String, branchId: String, orderAmount: Double? = null): JSONObject {
        val body = JSONObject()
            .put("code", code)
            .put("branchId", branchId)
        if (orderAmount != null) body.put("orderAmount", orderAmount)
        return post("/api/pos/coupons/validate", body.toString())
    }

    // ── Loyalty ──

    fun getLoyalty(phone: String): JSONObject {
        return get("/api/pos/loyalty?phone=$phone")
    }

    fun loyaltyTransaction(phone: String, type: String, points: Int, orderId: String? = null): JSONObject {
        val body = JSONObject()
            .put("phone", phone)
            .put("type", type)
            .put("points", points)
        if (!orderId.isNullOrBlank()) body.put("orderId", orderId)
        return post("/api/pos/loyalty", body.toString())
    }

    // ── KDS ──

    fun getKdsTickets(branchId: String, station: String? = null): JSONObject {
        val params = buildString {
            append("?branchId=$branchId")
            if (!station.isNullOrBlank()) append("&station=$station")
        }
        return get("/api/kds/tickets$params")
    }

    fun updateTicketStatus(ticketId: String, status: String): JSONObject {
        val body = JSONObject().put("status", status).toString()
        return patch("/api/kds/tickets/$ticketId/status", body)
    }

    // ── Waiter ──

    fun getWaiterCalls(branchId: String): JSONObject {
        return get("/api/waiter/calls?branchId=$branchId")
    }

    fun ackWaiterCall(callId: String): JSONObject {
        val body = JSONObject().put("actorUserId", prefs.userId ?: "android_pos").toString()
        return patch("/api/waiter/calls/$callId/ack", body)
    }

    fun resolveWaiterCall(callId: String): JSONObject {
        val body = JSONObject().put("actorUserId", prefs.userId ?: "android_pos").toString()
        return patch("/api/waiter/calls/$callId/resolve", body)
    }

    fun getReadyOrders(branchId: String): JSONObject {
        return get("/api/waiter/ready-queue?branchId=$branchId")
    }

    // ── Tables ──

    fun getTables(branchId: String): JSONObject {
        return get("/api/admin/branches/$branchId/tables")
    }

    // ── Sync ──

    fun syncEvents(branchId: String, deviceId: String, events: org.json.JSONArray): JSONObject {
        val body = JSONObject()
            .put("branchId", branchId)
            .put("deviceId", deviceId)
            .put("events", events)
            .toString()
        return post("/api/pos/sync/events", body, useDeviceAuth = true)
    }

    // ── Print ──

    fun printReceipt(orderId: String, printerType: String = "RECEIPT"): JSONObject {
        val body = JSONObject()
            .put("orderId", orderId)
            .put("printerType", printerType)
            .toString()
        return post("/api/pos/print/receipt", body)
    }

    // ── HTTP Methods ──

    private fun get(path: String, useDeviceAuth: Boolean = false): JSONObject {
        val request = Request.Builder()
            .url("${prefs.apiBaseUrl}$path")
            .get()
            .apply {
                if (useDeviceAuth) {
                    header("x-device-id", prefs.deviceId ?: "")
                    header("x-device-token", prefs.deviceToken ?: "")
                }
            }
            .build()
        return execute(request)
    }

    private fun post(path: String, body: String, auth: Boolean = true, useDeviceAuth: Boolean = false): JSONObject {
        val request = Request.Builder()
            .url("${prefs.apiBaseUrl}$path")
            .post(body.toRequestBody(jsonMediaType))
            .apply {
                if (useDeviceAuth) {
                    header("x-device-id", prefs.deviceId ?: "")
                    header("x-device-token", prefs.deviceToken ?: "")
                }
                if (!auth) header("X-No-Auth", "true")
            }
            .build()
        return execute(request)
    }

    private fun patch(path: String, body: String): JSONObject {
        val request = Request.Builder()
            .url("${prefs.apiBaseUrl}$path")
            .patch(body.toRequestBody(jsonMediaType))
            .build()
        return execute(request)
    }

    private fun execute(request: Request): JSONObject {
        return executeInternal(request, allowAutoRefresh = true)
    }

    private fun executeInternal(request: Request, allowAutoRefresh: Boolean): JSONObject {
        httpClient.newCall(request).execute().use { response ->
            val raw = response.body?.string() ?: "{}"
            val json = JSONObject(raw)

            if (response.isSuccessful) return json

            // Auto-refresh expired access tokens once, then retry the original request.
            val canRefresh =
                allowAutoRefresh &&
                    response.code == 401 &&
                    request.header("x-device-token") == null &&
                    request.header("X-No-Auth") == null &&
                    !prefs.refreshToken.isNullOrBlank()

            if (canRefresh && tryRefreshTokens()) {
                return executeInternal(request, allowAutoRefresh = false)
            }

            val errorMsg = json.optString("error", "Request failed: ${response.code}")
            error(errorMsg)
        }
    }

    private fun tryRefreshTokens(): Boolean {
        val refresh = prefs.refreshToken ?: return false
        val body = JSONObject()
            .put("refreshToken", refresh)
            .toString()
        val request = Request.Builder()
            .url("${prefs.apiBaseUrl}/api/auth/refresh")
            .post(body.toRequestBody(jsonMediaType))
            .header("X-No-Auth", "true")
            .build()

        return runCatching {
            executeInternal(request, allowAutoRefresh = false)
        }.mapCatching { result ->
            val data = result.optJSONObject("data") ?: return@mapCatching false
            val newAccess = data.optString("accessToken", "")
            val newRefresh = data.optString("refreshToken", "")
            if (newAccess.isBlank() || newRefresh.isBlank()) return@mapCatching false
            prefs.accessToken = newAccess
            prefs.refreshToken = newRefresh
            true
        }.getOrElse {
            prefs.clearAuth()
            false
        }
    }
}

/**
 * OkHttp interceptor that adds JWT Bearer token to all requests
 * (unless marked with X-No-Auth header or device auth headers present).
 */
private class AuthInterceptor(private val prefs: SecurePrefs) : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val original = chain.request()

        // Skip auth for explicitly marked requests
        if (original.header("X-No-Auth") != null) {
            val cleaned = original.newBuilder().removeHeader("X-No-Auth").build()
            return chain.proceed(cleaned)
        }

        // Skip if device auth headers are present
        if (original.header("x-device-token") != null) {
            return chain.proceed(original)
        }

        // Add JWT Bearer token
        val token = prefs.accessToken
        if (!token.isNullOrBlank()) {
            val authed = original.newBuilder()
                .header("Authorization", "Bearer $token")
                .header("x-org-id", prefs.organizationId)
                .header("x-branch-id", prefs.branchId)
                .build()
            return chain.proceed(authed)
        }

        return chain.proceed(original)
    }
}

File: ios/ios/Network/ApiClient.swift
import Foundation
import Combine

enum APIError: Error {
    case invalidURL
    case requestFailed(statusCode: Int, message: String?)
    case decodingError(Error)
    case unknown(Error)
}

extension APIError: LocalizedError {
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "رابط الخادم غير صالح"
        case .requestFailed(let statusCode, let message):
            let normalizedMessage = message?.lowercased() ?? ""
            if normalizedMessage.contains("jwt expired") {
                return "انتهت الجلسة. جاري تجديد الدخول أو أعد تسجيل الدخول."
            }
            if let message, !message.isEmpty {
                return "فشل الطلب (\(statusCode)): \(message)"
            }
            return "فشل الطلب برمز \(statusCode)"
        case .decodingError(let error):
            return "فشل قراءة البيانات: \(error.localizedDescription)"
        case .unknown(let error):
            return error.localizedDescription
        }
    }
}

class ApiClient: ObservableObject {
    let configStore: PosConfigStore
    private let session: URLSession

    init(configStore: PosConfigStore) {
        self.configStore = configStore
        self.session = URLSession.shared
    }

    private func baseURL() -> String {
        return configStore.apiBaseUrl
    }

    private func normalizedBaseURL() -> String {
        var raw = baseURL().trimmingCharacters(in: .whitespacesAndNewlines)

        if raw.isEmpty {
            raw = "http://localhost:3002"
        }

        if !(raw.hasPrefix("http://") || raw.hasPrefix("https://")) {
            raw = "http://\(raw)"
        }

        if raw.hasSuffix("/") {
            raw.removeLast()
        }

        return raw
    }

    private func createRequest(path: String, method: String, body: Data? = nil, useAuth: Bool = true, useDeviceAuth: Bool = false) -> URLRequest? {
        let fullUrlString = "\(normalizedBaseURL())\(path)"
        guard let url = URL(string: fullUrlString) else {
            print("Invalid URL in ApiClient.createRequest: \(fullUrlString)")
            return nil
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if useAuth, let token = configStore.accessToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            request.setValue(configStore.branchId, forHTTPHeaderField: "x-branch-id")
            request.setValue(configStore.organizationId, forHTTPHeaderField: "x-org-id")
        }

        if useDeviceAuth {
            if let deviceId = configStore.deviceId {
                request.setValue(deviceId, forHTTPHeaderField: "x-device-id")
            }
            if let deviceToken = configStore.deviceToken {
                request.setValue(deviceToken, forHTTPHeaderField: "x-device-token")
            }
        }

        request.httpBody = body
        return request
    }

    private func extractServerMessage(from data: Data) -> String? {
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let error = json["error"] as? String, !error.isEmpty {
                return error
            }
            if let message = json["message"] as? String, !message.isEmpty {
                return message
            }
        }

        if let raw = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty {
            return raw
        }

        return nil
    }

    private func shouldRefreshAuth(statusCode: Int, message: String?, request: URLRequest, isRetry: Bool) -> Bool {
        guard !isRetry,
              !(request.url?.path.contains("auth/refresh") ?? false),
              !(request.url?.path.contains("auth/login") ?? false),
              configStore.refreshToken != nil else {
            return false
        }

        if statusCode == 401 {
            return true
        }

        let normalized = (message ?? "").lowercased()
        return statusCode == 400 && (
            normalized.contains("jwt expired") ||
            normalized.contains("unauthorized") ||
            normalized.contains("invalid token")
        )
    }

    private func requestByRefreshingBearerToken(from request: URLRequest) async throws -> URLRequest {
        _ = try await refreshToken()
        var newRequest = request
        if let newToken = configStore.accessToken {
            newRequest.setValue("Bearer \(newToken)", forHTTPHeaderField: "Authorization")
        }
        return newRequest
    }

    private func performData(_ request: URLRequest, isRetry: Bool = false) async throws -> Data {
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.unknown(NSError(domain: "Invalid Response", code: 0))
        }

        if !(200...299).contains(httpResponse.statusCode) {
            let message = extractServerMessage(from: data)

            if shouldRefreshAuth(statusCode: httpResponse.statusCode, message: message, request: request, isRetry: isRetry) {
                do {
                    let newRequest = try await requestByRefreshingBearerToken(from: request)
                    return try await performData(newRequest, isRetry: true)
                } catch {
                    DispatchQueue.main.async {
                        self.configStore.accessToken = nil
                        self.configStore.refreshToken = nil
                    }
                }
            }

            throw APIError.requestFailed(statusCode: httpResponse.statusCode, message: message)
        }

        return data
    }

    private func performJSONObject(_ request: URLRequest, isRetry: Bool = false) async throws -> [String: Any] {
        let data = try await performData(request, isRetry: isRetry)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw APIError.decodingError(NSError(domain: "Invalid JSON", code: 0))
        }

        return json
    }

    private func perform<T: Decodable>(_ request: URLRequest, isRetry: Bool = false) async throws -> T {
        do {
            let data = try await performData(request, isRetry: isRetry)

            do {
                let decoder = JSONDecoder()
                return try decoder.decode(T.self, from: data)
            } catch {
                print("Decoding error: \(error)")
                throw APIError.decodingError(error)
            }
        } catch let error as APIError {
            throw error
        } catch {
            throw APIError.unknown(error)
        }
    }

    // MARK: - Auth

    func login(email: String, password: String) async throws -> LoginResponse {
        let body: [String: Any] = ["email": email, "password": password]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/auth/login", method: "POST", body: jsonData, useAuth: false) else {
            throw APIError.invalidURL
        }
        
        let wrapper: LoginResponseWrapper = try await perform(request)
        let response = wrapper.data
        
        // Auto-register device for demo to get full API access before triggering UI updates
        _ = try? await registerDevice(code: "IPAD-\(UUID().uuidString.prefix(4))")
        
        // Update tokens
        DispatchQueue.main.async {
            self.configStore.accessToken = response.accessToken
            self.configStore.refreshToken = response.refreshToken
        }
        return response
    }
    
    func refreshToken() async throws -> LoginResponse {
        guard let refreshToken = configStore.refreshToken else {
            throw APIError.unknown(NSError(domain: "No refresh token", code: 401))
        }
        
        let body: [String: Any] = ["refreshToken": refreshToken]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/auth/refresh", method: "POST", body: jsonData, useAuth: false) else {
            throw APIError.invalidURL
        }
        
        let wrapper: LoginResponseWrapper = try await perform(request)
        let response = wrapper.data
        DispatchQueue.main.async {
            self.configStore.accessToken = response.accessToken
            self.configStore.refreshToken = response.refreshToken
        }
        return response
    }
    
    func logout() async throws {
        guard let refreshToken = configStore.refreshToken else { return }
        
        let body: [String: Any] = ["refreshToken": refreshToken]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/auth/logout", method: "POST", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
        
        DispatchQueue.main.async {
            self.configStore.accessToken = nil
            self.configStore.refreshToken = nil
        }
    }

    // MARK: - Device Registration

    struct DeviceRegistrationResponse: Codable {
        let data: DeviceRegistrationData
    }

    struct DeviceRegistrationData: Codable {
        let deviceId: String
        let token: String
    }

    func registerDevice(code: String) async throws -> DeviceRegistrationData {
        let body: [String: Any] = [
            "branchId": configStore.branchId,
            "code": code,
            "name": "iPad POS",
            "platform": "IOS"
        ]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/pos/devices/register", method: "POST", body: jsonData, useAuth: false) else {
            throw APIError.invalidURL
        }

        let response: DeviceRegistrationResponse = try await perform(request)
        
        DispatchQueue.main.async {
            self.configStore.deviceId = response.data.deviceId
            self.configStore.deviceToken = response.data.token
        }
        
        return response.data
    }
    
    func rotateDeviceToken() async throws -> DeviceRegistrationData {
        guard let deviceId = configStore.deviceId,
              let oldToken = configStore.deviceToken else {
            throw APIError.unknown(NSError(domain: "Missing device credentials", code: 400))
        }
        
        let body: [String: Any] = [
            "deviceId": deviceId,
            "oldToken": oldToken
        ]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/pos/devices/rotate-token", method: "POST", body: jsonData, useAuth: false) else {
            throw APIError.invalidURL
        }
        
        let response: DeviceRegistrationResponse = try await perform(request)
        
        DispatchQueue.main.async {
            self.configStore.deviceToken = response.data.token
        }
        
        return response.data
    }

    // MARK: - Snapshot

    struct SnapshotResponse: Codable {
        let data: SnapshotData
    }
    
    // Custom decoding for Snapshot because the response structure might be nested
    // Looking at PosApiClient.kt: `val data = root.optJSONObject("data")`
    // So it returns { data: { ... } }

    func fetchSnapshot() async throws -> SnapshotData {
        guard let request = createRequest(path: "/api/pos/sync/snapshot?branchId=\(configStore.branchId)", method: "GET", useAuth: true, useDeviceAuth: true) else {
            throw APIError.invalidURL
        }
        
        let response: SnapshotResponse = try await perform(request)
        return response.data
    }

    // MARK: - Tables

    struct TablesResponse: Codable {
        let data: TablesData
    }

    struct TablesData: Codable {
        let tables: [TableDto]
        let areas: [TableAreaDto]
    }

    func getTables() async throws -> TablesData {
        guard let request = createRequest(path: "/api/admin/branches/\(configStore.branchId)/tables", method: "GET", useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let response: TablesResponse = try await perform(request)
        return response.data
    }

    // MARK: - Orders

    struct OrdersResponse: Decodable {
        let data: OrdersData
    }

    struct OrdersData: Decodable {
        let orders: [OrderDto]
        let pagination: Pagination?
    }

    struct Pagination: Decodable {
        let total: Int
        let totalPages: Int
    }

    func getOrders(page: Int = 1, limit: Int = 50, status: String? = nil) async throws -> OrdersData {
        var path = "/api/orders?branchId=\(configStore.branchId)&page=\(page)&limit=\(limit)"
        if let status = status {
            path += "&status=\(status)"
        }
        
        guard let request = createRequest(path: path, method: "GET", useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let response: OrdersResponse = try await perform(request)
        return response.data
    }
    
    func voidOrder(orderId: String, reason: String? = nil) async throws {
        var body: [String: Any] = [:]
        if let reason = reason {
            body["reason"] = reason
        }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/orders/\(orderId)/void", method: "PATCH", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
    }
    
    func refundOrder(orderId: String, paymentId: String, amount: Double, reason: String? = nil) async throws {
        var body: [String: Any] = [
            "paymentId": paymentId,
            "amount": amount
        ]
        if let reason = reason {
            body["reason"] = reason
        }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/orders/\(orderId)/refund", method: "PATCH", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
    }
    
    func applyDiscount(orderId: String, discountAmount: Double, reason: String? = nil) async throws {
        var body: [String: Any] = [
            "discountAmount": discountAmount
        ]
        if let reason = reason {
            body["reason"] = reason
        }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/orders/\(orderId)/discount", method: "PATCH", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
    }
    
    func createOrder(type: String, items: [[String: Any]], tableId: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [
            "branchId": configStore.branchId,
            "type": type,
            "items": items
        ]
        if let tableId = tableId {
            body["tableId"] = tableId
        }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/orders", method: "POST", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    func updateOrderStatus(orderId: String, status: String) async throws {
        let body: [String: Any] = ["status": status]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/orders/\(orderId)/status", method: "PATCH", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
    }
    
    func splitOrder(orderId: String, items: [[String: Any]]) async throws -> [String: Any] {
        let body: [String: Any] = ["items": items]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/orders/\(orderId)/split", method: "POST", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    func mergeOrders(sourceOrderId: String, targetOrderId: String) async throws -> [String: Any] {
        let body: [String: Any] = ["targetOrderId": targetOrderId]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/orders/\(sourceOrderId)/merge", method: "POST", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    func transferTable(orderId: String, tableId: String, reason: String? = nil) async throws {
        var body: [String: Any] = ["tableId": tableId]
        if let reason = reason {
            body["reason"] = reason
        }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/orders/\(orderId)/transfer-table", method: "PATCH", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
    }
    
    func getInvoice(orderId: String) async throws -> Data? {
        guard let request = createRequest(path: "/api/orders/\(orderId)/invoice", method: "GET", useAuth: true) else {
            throw APIError.invalidURL
        }

        return try? await performData(request)
    }
    
    // MARK: - Coupons
    
    func validateCoupon(code: String, orderAmount: Double? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [
            "code": code,
            "branchId": configStore.branchId
        ]
        if let orderAmount = orderAmount {
            body["orderAmount"] = orderAmount
        }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/pos/coupons/validate", method: "POST", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    // MARK: - Loyalty
    
    func getLoyalty(phone: String) async throws -> [String: Any] {
        guard let request = createRequest(path: "/api/pos/loyalty?phone=\(phone)", method: "GET", useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    func loyaltyTransaction(phone: String, type: String, points: Int, orderId: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [
            "phone": phone,
            "type": type,
            "points": points
        ]
        if let orderId = orderId {
            body["orderId"] = orderId
        }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/pos/loyalty", method: "POST", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    // MARK: - Sync
    
    func syncEvents(events: [[String: Any]]) async throws -> [String: Any] {
        guard let deviceId = configStore.deviceId else {
            throw APIError.unknown(NSError(domain: "Missing device ID", code: 400))
        }
        
        let body: [String: Any] = [
            "branchId": configStore.branchId,
            "deviceId": deviceId,
            "events": events
        ]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/pos/sync/events", method: "POST", body: jsonData, useAuth: true, useDeviceAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    // MARK: - Print
    
    func printReceipt(orderId: String, printerType: String = "RECEIPT", copies: Int = 1, deviceId: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [
            "orderId": orderId,
            "printerType": printerType,
            "copies": max(1, copies)
        ]
        if let deviceId {
            body["deviceId"] = deviceId
        }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/pos/print/receipt", method: "POST", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    // MARK: - KDS
    
    func getKdsTickets() async throws -> [String: Any] {
        guard let request = createRequest(path: "/api/kds/tickets?branchId=\(configStore.branchId)", method: "GET", useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    func updateTicketStatus(ticketId: String, status: String) async throws {
        let body: [String: Any] = ["status": status]
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        
        guard let request = createRequest(path: "/api/kds/tickets/\(ticketId)/status", method: "PATCH", body: jsonData, useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
    }
    
    // MARK: - Waiter Calls
    
    func getWaiterCalls() async throws -> [String: Any] {
        guard let request = createRequest(path: "/api/waiter/calls?branchId=\(configStore.branchId)", method: "GET", useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    func getReadyOrders() async throws -> [String: Any] {
        guard let request = createRequest(path: "/api/waiter/ready-queue?branchId=\(configStore.branchId)", method: "GET", useAuth: true) else {
            throw APIError.invalidURL
        }

        return try await performJSONObject(request)
    }
    
    func ackWaiterCall(callId: String) async throws {
        guard let request = createRequest(path: "/api/waiter/calls/\(callId)/ack", method: "PATCH", useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
    }
    
    func resolveWaiterCall(callId: String) async throws {
        guard let request = createRequest(path: "/api/waiter/calls/\(callId)/resolve", method: "PATCH", useAuth: true) else {
            throw APIError.invalidURL
        }
        
        let _: EmptyResponse = try await perform(request)
    }
    
    struct EmptyResponse: Codable {}
}
